compiler/rustc_middle/src/ty/layout.rs

```diff
@@ -2097,42 +2103,44 @@ where
             }
 
             // Arrays and slices.
-            ty::Array(element, _) | ty::Slice(element) => element,
-            ty::Str => tcx.types.u8,
+            ty::Array(element, _) | ty::Slice(element) => Ok(element),
+            ty::Str => Ok(tcx.types.u8),
 
             // Tuples, generators and closures.
-            ty::Closure(_, ref substs) => Self::field(substs.as_closure().tupled_upvars_ty()),
+            ty::Closure(_, ref substs) => {
+                Self::ty_and_layout_kind(this, cx, i, substs.as_closure().tupled_upvars_ty())
+            }
 
             ty::Generator(def_id, ref substs, _) => match this.variants {
-                Variants::Single { index } => substs
+                Variants::Single { index } => Ok(substs
                     .as_generator()
                     .state_tys(def_id, tcx)
                     .nth(index.as_usize())
                     .unwrap()
                     .nth(i)
-                    .unwrap(),
+                    .unwrap()),
                 Variants::Multiple { ref tag, tag_field, .. } => {
                     if i == tag_field {
-                        return tag_layout(tag);
+                        return Err(tag_layout(tag));
                     }
-                    substs.as_generator().prefix_tys().nth(i).unwrap()
+                    Ok(substs.as_generator().prefix_tys().nth(i).unwrap())
                 }
             },
 
-            ty::Tuple(tys) => tys[i].expect_ty(),
+            ty::Tuple(tys) => Ok(tys[i].expect_ty()),
 
             // SIMD vector types.
-            ty::Adt(def, ..) if def.repr.simd() => this.ty.simd_type(tcx),
+            ty::Adt(def, ..) if def.repr.simd() => Ok(ty.simd_type(tcx)),
 
             // ADTs.
             ty::Adt(def, substs) => {
                 match this.variants {
-                    Variants::Single { index } => def.variants[index].fields[i].ty(tcx, substs),
+                    Variants::Single { index } => Ok(def.variants[index].fields[i].ty(tcx, substs)),
 
                     // Discriminant field for enums (where applicable).
                     Variants::Multiple { ref tag, .. } => {
                         assert_eq!(i, 0);
-                        return tag_layout(tag);
+                        return Err(tag_layout(tag));
```

Comment:
I don't think this is necessarily an err ...

Reply:
I am using the Err not because it is an error, but to know whether to return or to pass as parameter the output of `ty_and_layout_kind` into `layout_of`

Reply:
Nvm, I see what you mean now: https://github.com/rust-lang/rust/pull/78725#discussion_r522574962